### PR TITLE
Improvements on Token Scoping

### PR DESCRIPTION
**Pull request by peterthomassen**
- [x] Change policy `id` field to UUID
- [x] Add columns `subname`, `type` to policies
- [x] Adjust policies view for new fields
- [x] Include subname/type in Token default policies
- [x] Extend RRset serializers to honor policies
- [x] Update docs
- [ ] ~~Turn `perm_manage_tokens` into a virtual field (`true` iff token has no policies) -- rethink: no policy means "allow/create delete domains", but not necessarily manage tokens~~

Related: #347

### CODE REVIEW

File: api/desecapi/models/tokens.py [hunk 1]
```diff
@@ -146,6 +148,16 @@ class TokenDomainPolicy(ExportModelOperationsMixin("TokenDomainPolicy"), models.
...
     domain = models.ForeignKey("Domain", on_delete=models.CASCADE, null=True)
+    # TODO use RRset.subname_validators etc?
+    subname = models.CharField(
+        max_length=178,
+        blank=True,
+        null=True,
+        validators=RRset.subname.field._validators,
+    )
+    type = models.CharField(
+        max_length=10, null=True, validators=RRset.type.field._validators
```
Comment (nils-wisiol) [inline api/desecapi/models/tokens.py:122]: make sure that effective validators are equal for `TokenDomainPolicy` and `RRset` `subname` and `type`, e.g. by using Django console.

File: api/desecapi/tests/test_token_domain_policy.py [hunk 2]
```diff
@@ -142,91 +151,117 @@ def test_policy_lifecycle(self):
...
-        self.assertEqual(response.data, self.default_data | data)
+        self.assertEqual(response.data, self.default_data | data | {"id": policy_id})
 
         ## can't create policy for other user's domain
-        data = dict(domain=self.other_domain.name, perm_dyndns=True, perm_write=True)
+        data = {
+            "domain": self.other_domain.name,
+            "subname": None,
+            "type": None,
+            "perm_write": True,
```
Comment (nils-wisiol) [inline api/desecapi/tests/test_token_domain_policy.py:266]: perm_dyndns missing

File: api/desecapi/tests/test_token_domain_policy.py [hunk 3]
```diff
@@ -142,91 +151,117 @@ def test_policy_lifecycle(self):
...
         self.assertEqual(response.data["domain"][0].code, "does_not_exist")
 
         ## another policy
-        data = dict(domain=self.my_domains[0].name, perm_dyndns=True)
+        data = {
+            "domain": self.my_domains[0].name,
+            "subname": None,
+            "type": None,
+            "perm_dyndns": True,
+            "perm_write": True,
```
Comment (nils-wisiol) [inline api/desecapi/tests/test_token_domain_policy.py:279]: consider removing `perm_write`

File: api/desecapi/tests/test_token_domain_policy.py [hunk 4]
```diff
@@ -142,91 +151,117 @@ def test_policy_lifecycle(self):
...
         self.assertStatus(response, status.HTTP_200_OK)
-        self.assertEqual(response.data, self.default_data | data)
+        self.assertEqual(response.data, self.default_data | data | {"id": policy_id})
 
         # List: now has two elements
         response = self.client.list_policies(self.token, using=self.token_manage)
         self.assertStatus(response, status.HTTP_200_OK)
         self.assertEqual(len(response.data), 2)
 
+        default_policy_id = None
```
Comment (nils-wisiol) [inline api/desecapi/tests/test_token_domain_policy.py]: record id when creating default policy

File: api/desecapi/tests/test_token_domain_policy.py [hunk 5]
```diff
@@ -142,91 +151,117 @@ def test_policy_lifecycle(self):
...
+                default_policy_id = policy["id"]
+                break
+
         # Change
         ## all fields of a policy
-        data = dict(domain=self.my_domains[1].name, perm_dyndns=False, perm_write=True)
+        data = dict(
+            domain=self.my_domains[1].name,
+            subname=None,
+            type=None,
```
Comment (nils-wisiol) [inline api/desecapi/tests/test_token_domain_policy.py]: also change subname/type (or change comment in l234)

File: api/desecapi/models/tokens.py [hunk 6]
```diff
@@ -89,61 +89,24 @@ def generate_key(self):
...
+            )
+            for field in ["domain", "subname", "type"]
+        ]
         return (
-            self.tokendomainpolicy_set.filter(Q(domain=domain) | Q(domain__isnull=True))
-            .order_by(order_by)
+            self.tokendomainpolicy_set.filter(
+                Q(domain=domain) | Q(domain__isnull=True),
+                Q(subname=subname) | Q(subname__isnull=True),
+                Q(type=domain) | Q(type__isnull=True),
```
Comment (nils-wisiol) [inline api/desecapi/models/tokens.py]: change to `type=type`, make sure this is tested

File: api/desecapi/models/tokens.py [hunk 7]
```diff
@@ -167,42 +130,68 @@ class TokenDomainPolicy(ExportModelOperationsMixin("TokenDomainPolicy"), models.
...
-                        "domain": "Policy precedence: Cannot disable default policy when others exist."
-                    }
+                    "Policy precedence: The first policy must be the default policy."
                 )
-        else:  # create
-            # Can't violate policy precedence (default policy has to be first)
-            if (self.domain is not None) and (default_policy is None):
+        else:  # update
+            # Can't change policy's default status ("domain NULLness") to maintain policy precedence
+            if self.has_default_policy_settings != (self.pk == default_policy.pk):
```
Comment (nils-wisiol) [inline api/desecapi/models/tokens.py:181]:
```
is_default_policy_when_loaded = self.pk == default_policy.pk
is_default_policy_after_saving = self.has_default_policy_settings
# Can't make non-default policy default, can't make default policy non-default
if is_default_policy_when_loaded != is_default_policy_after_saving:
```

Comment (nils-wisiol) [reply api/desecapi/models/tokens.py:181]: Also update error message to catch both cases

File: api/desecapi/serializers/tokens.py [hunk 8]
```diff
@@ -73,4 +73,4 @@ def save(self, **kwargs):
         try:
             return super().save(**kwargs)
         except django.core.exceptions.ValidationError as exc:
-            raise serializers.ValidationError(exc.message_dict, code="precedence")
+            raise serializers.ValidationError(exc.message, code="precedence")
```
Comment (nils-wisiol) [inline api/desecapi/serializers/tokens.py:75]: undo to keep compatibility and prepare for usage in webapp

File: api/desecapi/views/tokens.py [hunk 9]
```diff
@@ -82,4 +82,4 @@ def perform_destroy(self, instance):
         try:
             super().perform_destroy(instance)
         except django.core.exceptions.ValidationError as exc:
-            raise ValidationError(exc.message_dict, code="precedence")
+            raise ValidationError(exc.message, code="precedence")
```
Comment (nils-wisiol) [inline api/desecapi/views/tokens.py:91]: undo?

File: api/desecapi/permissions.py [hunk 10]
```diff
@@ -80,50 +82,22 @@ class TokenNoDomainPolicy(permissions.BasePermission):
     """
```
Comment (nils-wisiol) [inline api/desecapi/permissions.py:82]: update comment, remove 'domain'

File: api/desecapi/tests/test_token_domain_policy.py (hunk 2 above)
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py:266]: done

File: api/desecapi/tests/test_token_domain_policy.py (hunk 3 above)
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py:279]: done

File: api/desecapi/tests/test_token_domain_policy.py (hunk 4 above)
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py]: done

File: api/desecapi/tests/test_token_domain_policy.py (hunk 5 above)
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py]: done

File: api/desecapi/models/tokens.py (hunk 6 above)
Comment (peterthomassen) [reply api/desecapi/models/tokens.py]: done

File: api/desecapi/models/tokens.py (hunk 7 above)
Comment (peterthomassen) [reply api/desecapi/models/tokens.py:181]: done

File: api/desecapi/models/tokens.py [hunk 11]
```diff
@@ -167,42 +130,68 @@ class TokenDomainPolicy(ExportModelOperationsMixin("TokenDomainPolicy"), models.
...
 
     def delete(self, *args, **kwargs):
         # Can't delete default policy when others exist
-        if (self.domain is None) and self.token.tokendomainpolicy_set.exclude(
-            domain__isnull=True
-        ).exists():
+        if (
+            self.has_default_policy_settings
+            and self.token.tokendomainpolicy_set.exclude(pk=self.pk).exists()
+        ):
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models/tokens.py:191]: done

File: api/desecapi/tests/test_token_domain_policy.py [hunk 12]
```diff
@@ -57,6 +57,79 @@ def setUp(self):
         self.token_manage = self.create_token(self.owner, perm_manage_tokens=True)
         self.other_token = self.create_token(self.user)
 
+    def test_get_policy(self):
+        def get_policy(domain, subname, type):
+            return self.token.get_policy(
+                models.RRset(domain=domain, subname=subname, type=type)
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py:64]: done

File: api/desecapi/tests/test_token_domain_policy.py [hunk 13]
```diff
@@ -498,11 +552,6 @@ def test_domain_owner_equals_token_user(self):
                     type=None,
                 ).save()
 
-        self.token.user = self.other_domain.owner
-        with self.assertRaises(IntegrityError):
-            with transaction.atomic():  # https://stackoverflow.com/a/23326971/6867099
-                self.token.save()
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/tests/test_token_domain_policy.py:447]: restored

File: api/desecapi/serializers/tokens.py (hunk 8 above)
Comment (peterthomassen) [reply api/desecapi/serializers/tokens.py:75]: done

File: api/desecapi/permissions.py (hunk 10 above)
Comment (peterthomassen) [reply api/desecapi/permissions.py:82]: done

File: api/desecapi/models/tokens.py (hunk 1 above)
Comment (peterthomassen) [reply api/desecapi/models/tokens.py:122]: done

**Pull request merged**